pkg/ocicni/ocicni.go

```diff
@@ -174,24 +169,102 @@ func (plugin *cniNetworkPlugin) Name() string {
 	return CNIPluginName
 }
 
-func (plugin *cniNetworkPlugin) SetUpPod(netnsPath string, namespace string, name string, id string) error {
+func (plugin *cniNetworkPlugin) monitorNetDir(netnsPath string, namespace string, name string, id string, timeout time.Duration) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logrus.Errorf("could not create new watcher", err)
+		return err
+	}
+	defer watcher.Close()
+
+	monitorTimeout := make(chan bool)
+	monitorDone := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				logrus.Debugf("CNI monitoring event %v", event)
+				if event.Op&fsnotify.Create != fsnotify.Create {
+					continue
+				}
+
+				if err := plugin.setUpPod(netnsPath, namespace, name, id); err != nil {
+					logrus.Errorf("could not set up pod with new CNI conf %v", err)
+				}
+
+				logrus.Debugf("CNI asynchronous setting succeeded")
+				close(monitorDone)
+				return
+
+			case err := <-watcher.Errors:
+				logrus.Errorf("CNI monitoring error %v", err)
+				close(monitorDone)
+				return
+
+			case <-monitorTimeout:
+				logrus.Errorf("CNI monitoring timeout")
+				return
+			}
+		}
+	}()
+
+	if err = watcher.Add(plugin.pluginDir); err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	select {
+	case <-monitorDone:
+		logrus.Debugf("CNI monitoring finished")
+
+	case <-time.After(time.Second * timeout):
+		close(monitorTimeout)
+	}
+
+	return nil
+}
+
+func (plugin *cniNetworkPlugin) setUpPod(netnsPath string, namespace string, name string, id string) error {
+	plugin.syncNetworkConfig()
+
 	if err := plugin.checkInitialized(); err != nil {
 		return err
 	}
 
 	_, err := plugin.loNetwork.addToNetwork(name, namespace, id, netnsPath)
 	if err != nil {
-		glog.Errorf("Error while adding to cni lo network: %s", err)
+		logrus.Errorf("Error while adding to cni lo network: %s", err)
 		return err
 	}
 
 	_, err = plugin.getDefaultNetwork().addToNetwork(name, namespace, id, netnsPath)
 	if err != nil {
-		glog.Errorf("Error while adding to cni network: %s", err)
+		logrus.Errorf("Error while adding to cni network: %s", err)
 		return err
 	}
 
 	return err
+
+}
+
+func (plugin *cniNetworkPlugin) SetUpPod(netnsPath string, namespace string, name string, id string) error {
+	// First let's sync with the latest configuration files
+	plugin.syncNetworkConfig()
+
+	// Now we can check if we really have a default network
+	if err := plugin.checkInitialized(); err != nil {
+		if err == errMissingDefaultNetwork {
+			// We are missing a default network.
+			// Let's wait 30s and see if a new configuration file appears.
+			go plugin.monitorNetDir(netnsPath, namespace, name, id, 30)
```

Comment:
@rajatchopra A go routine. I improved the PR a bit so that `TearDown()` kills that routine if it's still running.